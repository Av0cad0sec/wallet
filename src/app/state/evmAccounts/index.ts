import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { EvmAccounts } from './types'

export const initialState: EvmAccounts = {}

const slice = createSlice({
  name: 'evmAccounts',
  initialState,
  reducers: {
  },
})

export const { actions: evmAccountsActions } = slice

export default slice.reducer
