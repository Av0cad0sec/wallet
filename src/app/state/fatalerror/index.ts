import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { useInjectReducer } from 'utils/redux-injectors'
import { FatalErrorPayload, FatalErrorState } from './types'

export const initialState: FatalErrorState = {}

const slice = createSlice({
  name: 'fatalError',
  initialState,
  reducers: {
    setError(state, action: PayloadAction<FatalErrorPayload>) {
      if (!state.error) {
        state.error = action.payload
      }
    },
  },
})

export const { actions: fatalErrorActions } = slice

export const useFatalErrorSlice = () => {
  useInjectReducer({ key: slice.name, reducer: slice.reducer })
  return { actions: slice.actions }
}
