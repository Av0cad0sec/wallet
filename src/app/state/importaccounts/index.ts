import { PayloadAction } from '@reduxjs/toolkit'
import { ErrorPayload } from 'types/errors'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { ImportAccountsListAccount, ImportAccountsState, ImportAccountsStep } from './types'

export const initialState: ImportAccountsState = {
  accounts: [],
  showAccountsSelectionModal: false,
  step: ImportAccountsStep.Idle,
}

const slice = createSlice({
  name: 'importAccounts',
  initialState,
  reducers: {
    clear(state) {
      state.accounts = []
      state.error = undefined
      state.step = ImportAccountsStep.Idle
      state.showAccountsSelectionModal = false
    },
    enumerateAccountsFromLedger(state) {
      state.accounts = []
      state.showAccountsSelectionModal = true
      state.step = ImportAccountsStep.Idle
    },
    enumerateAccountsFromMnemonic(state, _action: PayloadAction<string>) {
      state.step = ImportAccountsStep.Idle
      state.accounts = []
      state.showAccountsSelectionModal = true
    },
    toggleAccount(state, action: PayloadAction<number>) {
      const index = action.payload
      state.accounts[index].selected = !state.accounts[index].selected
    },
    accountsListed(state, action: PayloadAction<ImportAccountsListAccount[]>) {
      state.accounts = action.payload
    },
    setStep(state, action: PayloadAction<ImportAccountsStep>) {
      state.step = action.payload
    },
    operationFailed(state, action: PayloadAction<ErrorPayload>) {
      state.error = action.payload
      state.step = ImportAccountsStep.Idle
    },
  },
})

export const { actions: importAccountsActions } = slice

export default slice.reducer
