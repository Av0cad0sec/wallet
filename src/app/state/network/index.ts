import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { useInjectReducer, useInjectSaga } from 'utils/redux-injectors'
import { networkSaga } from './saga'
import { NetworkState, NetworkType } from './types'

export const initialState: NetworkState = {
  ticker: '',
  chainContext: '',
  selectedNetwork: 'local',
}

const slice = createSlice({
  name: 'network',
  initialState,
  reducers: {
    selectNetwork(state, action: PayloadAction<NetworkType>) {},
    networkSelected(state, action: PayloadAction<NetworkState>) {
      Object.assign(state, action.payload)
    },
  },
})

export const { actions: networkActions } = slice

export const useNetworkSlice = () => {
  useInjectReducer({ key: slice.name, reducer: slice.reducer })
  useInjectSaga({ key: slice.name, saga: networkSaga })
  return { actions: slice.actions }
}
