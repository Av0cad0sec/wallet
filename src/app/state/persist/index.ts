import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { PersistState } from './types'
import { runtimeIs } from 'config'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { createPersistedRootReducer } from 'store/reducers'
import { RootState } from 'types'

export const STORAGE_FIELD = 'oasis_wallet_persist_v1'

/** Syncing tabs is only needed in web app, not in extension. */
export const needsSyncingTabs = runtimeIs === 'webapp'

// Simulate with `delete window.BroadcastChannel`
export const isSyncingTabsSupported = typeof BroadcastChannel === 'function'

export function getInitialState(): PersistState {
  return {
    hasPersistedProfiles: !!localStorage.getItem(STORAGE_FIELD),
    // Disable persistence if tabs would override each other.
    isPersistenceUnsupported: needsSyncingTabs && !isSyncingTabsSupported,
    loading: false,
    stringifiedEncryptionKey: undefined,
    enteredWrongPassword: false,
  }
}

const persistSlice = createSlice({
  name: 'persist',
  initialState: getInitialState(),
  reducers: {
    setWrongPassword(state) {
      state.enteredWrongPassword = true
      state.loading = false
    },
    setUnlockedRootState(state, action: PayloadAction<{ rootState: RootState }>) {
      /**
       * Handled in {@link createPersistedRootReducer}.
       * Sets `state.loading = false` and `state.stringifiedEncryptionKey`.
       */
      return
    },

    // Handled in saga
    // ---------------
    /**
     * Set a password to encrypt state with and begin storing it to localStorage.
     */
    setPasswordAsync(state, action: PayloadAction<{ password: string }>) {
      state.loading = true
    },
    /**
     * Load encrypted state from localStorage and decrypt, or set {@link PersistState.enteredWrongPassword}.
     */
    unlockAsync(state, action: PayloadAction<{ password: string }>) {
      state.loading = true
    },
    /**
     * Remove everything from state, especially {@link PersistState.stringifiedEncryptionKey}.
     * UnlockGate should stop any components relying on missing state.
     */
    lockAsync(state) {
      state.loading = true
    },
    /**
     * Remove encrypted state from localStorage and reload.
     */
    eraseAsync(state) {
      state.loading = true
    },
  },
})

export const persistActions = persistSlice.actions

export default persistSlice.reducer
