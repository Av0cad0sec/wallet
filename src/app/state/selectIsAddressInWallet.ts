import { createSelector } from '@reduxjs/toolkit'
import { selectAccountAddress } from 'app/state/account/selectors'
import { selectAddress, selectIsOpen } from 'app/state/wallet/selectors'

export const selectIsAddressInWallet = createSelector(
  [selectIsOpen, selectAddress, selectAccountAddress],
  (walletIsOpen, walletAddress, address) => walletIsOpen && walletAddress === address,
)
