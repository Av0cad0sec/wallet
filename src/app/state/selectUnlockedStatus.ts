import { createSelector } from '@reduxjs/toolkit'
import { selectSlice as selectPersistSlice } from 'app/state/persist/selectors'
import { selectIsOpen } from 'app/state/wallet/selectors'

export const selectUnlockedStatus = createSelector([selectPersistSlice, selectIsOpen], (state, isOpen) => {
  if (state.hasPersistedProfiles) {
    if (state.stringifiedEncryptionKey === 'skipped') {
      if (isOpen) return 'openSkippedUnlockingProfile'
      return 'emptySkippedUnlockingProfile'
    }
    if (state.stringifiedEncryptionKey) return 'unlockedProfile'
    return 'lockedProfile'
  } else {
    if (isOpen) return 'openUnpersisted'
    return 'emptyUnpersisted'
  }
})
