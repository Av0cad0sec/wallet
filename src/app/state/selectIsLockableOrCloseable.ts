import { createSelector } from '@reduxjs/toolkit'
import { selectUnlockedStatus } from 'app/state/selectUnlockedStatus'
import { ExhaustedTypeError } from 'types/errors'

export const selectIsLockableOrCloseable = createSelector([selectUnlockedStatus], unlockedStatus => {
  if (unlockedStatus === 'skippedUnlockingProfile') return 'closeable'
  if (unlockedStatus === 'unlockedProfile') return 'lockable'
  if (unlockedStatus === 'openUnpersisted') return 'closeable'
  if (unlockedStatus === 'lockedProfile') return false
  if (unlockedStatus === 'emptyUnpersisted') return false
  throw new ExhaustedTypeError('Invalid unlocked status', unlockedStatus)
})
