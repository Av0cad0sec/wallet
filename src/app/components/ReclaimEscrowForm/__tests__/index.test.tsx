import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { transactionActions } from 'app/state/transaction'
import * as React from 'react'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'

import { ReclaimEscrowForm } from '..'

const renderComponent = (store: any, address: string, maxAmount: string, shares: string) =>
  render(
    <Provider store={store}>
      <ReclaimEscrowForm address={address} maxAmount={maxAmount} shares={shares} />
    </Provider>,
  )

describe('<ReclaimEscrowForm />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore()
    jest.resetAllMocks()
  })

  it('should match snapshot', () => {
    const component = renderComponent(store, 'dummy-address', '1000', '1000')
    expect(component.container.firstChild).toMatchSnapshot()
  })

  it('should error without an open wallet', () => {
    renderComponent(store, 'dummy-address', '1000000000000', '1000000000000')

    userEvent.type(screen.getByTestId('amount'), '1000')
    userEvent.click(screen.getByRole('button', { name: 'account.reclaimEscrow.reclaim' }))

    expect(screen.getByText('errors.noOpenWallet')).toBeInTheDocument()
  })

  it('should display the number of shares', () => {
    renderComponent(store, 'dummy-address', '1000000000000', '1000000000000')

    expect(screen.queryByTestId('numberOfShares')).toBeNull()
    userEvent.type(screen.getByTestId('amount'), '500')
    expect(screen.getByTestId('numberOfShares')).toBeInTheDocument()
  })

  it('should submit the transaction', () => {
    const spy = jest.spyOn(store, 'dispatch')
    renderComponent(store, 'dummy-address', '2000000000000', '1000000000000')

    userEvent.type(screen.getByTestId('amount'), '500')
    userEvent.click(screen.getByRole('button', { name: 'account.reclaimEscrow.reclaim' }))

    expect(spy).toHaveBeenCalledWith({
      payload: {
        amount: '500000000000',
        shares: '250000000000',
        type: 'reclaimEscrow',
        validator: 'dummy-address',
      },
      type: 'transaction/reclaimEscrow',
    } as ReturnType<typeof transactionActions.reclaimEscrow>)
  })

  it('reclaim all should submit the transaction', () => {
    const spy = jest.spyOn(store, 'dispatch')
    renderComponent(store, 'dummy-address', '2000000000000', '1000000000000')
    userEvent.click(screen.getByRole('button', { name: 'account.reclaimEscrow.reclaimAll' }))

    expect(spy).toHaveBeenCalledWith({
      payload: {
        amount: '2000000000000',
        shares: '1000000000000',
        type: 'reclaimEscrow',
        validator: 'dummy-address',
      },
      type: 'transaction/reclaimEscrow',
    } as ReturnType<typeof transactionActions.reclaimEscrow>)
  })

  it('reclaim all should not lose precision', () => {
    const spy = jest.spyOn(store, 'dispatch')
    renderComponent(store, 'dummy-address', '20000000000000002', '10000000000000001')
    userEvent.click(screen.getByRole('button', { name: 'account.reclaimEscrow.reclaimAll' }))

    expect(spy).toHaveBeenCalledWith({
      payload: {
        amount: '20000000000000002',
        shares: '10000000000000001',
        type: 'reclaimEscrow',
        validator: 'dummy-address',
      },
      type: 'transaction/reclaimEscrow',
    })
  })
})
