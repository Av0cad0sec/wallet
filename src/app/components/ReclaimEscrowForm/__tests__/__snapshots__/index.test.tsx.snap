// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReclaimEscrowForm /> should match snapshot 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-top: 12px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
}

.c9 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 24px;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  height: 24px;
}

.c8 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0092f6;
  border-radius: 4px;
  color: #16213E;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #0092f6;
  color: #F8F8F8;
  border-radius: 4px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
  height: 100%;
  max-width: none;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font-weight: bold;
}

.c8:hover {
  box-shadow: 0px 0px 0px 2px #0092f6;
}

.c8:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c8:focus > circle,
.c8:focus > ellipse,
.c8:focus > line,
.c8:focus > path,
.c8:focus > polygon,
.c8:focus > polyline,
.c8:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c8:focus::-moz-focus-inner {
  border: 0;
}

.c8:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c8:focus:not(:focus-visible) > circle,.c8:focus:not(:focus-visible) > ellipse,
.c8:focus:not(:focus-visible) > line,.c8:focus:not(:focus-visible) > path,
.c8:focus:not(:focus-visible) > polygon,.c8:focus:not(:focus-visible) > polyline,
.c8:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c8:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c10 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0092f6;
  border-radius: 4px;
  color: #16213E;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
  height: 100%;
  max-width: none;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font-weight: bold;
}

.c10:hover {
  box-shadow: 0px 0px 0px 2px #0092f6;
}

.c10:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus > circle,
.c10:focus > ellipse,
.c10:focus > line,
.c10:focus > path,
.c10:focus > polygon,
.c10:focus > polyline,
.c10:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus::-moz-focus-inner {
  border: 0;
}

.c10:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible) > circle,.c10:focus:not(:focus-visible) > ellipse,
.c10:focus:not(:focus-visible) > line,.c10:focus:not(:focus-visible) > path,
.c10:focus:not(:focus-visible) > polygon,.c10:focus:not(:focus-visible) > polyline,
.c10:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c4 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  font-size: 18px;
  line-height: 24px;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
}

.c4:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c4:focus > circle,
.c4:focus > ellipse,
.c4:focus > line,
.c4:focus > path,
.c4:focus > polygon,
.c4:focus > polyline,
.c4:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c4:focus::-moz-focus-inner {
  border: 0;
}

.c4::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c4::-moz-placeholder {
  color: #AAAAAA;
}

.c4:-ms-input-placeholder {
  color: #AAAAAA;
}

.c4::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c4::-moz-focus-inner {
  border: none;
  outline: none;
}

.c4:-moz-placeholder,
.c4::-moz-placeholder {
  opacity: 1;
}

.c3 {
  position: relative;
  width: 100%;
}

.c0 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #16213E;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c1 {
    padding-top: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c9 {
    width: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c5 {
    height: 12px;
  }
}

<div
  class="c0"
  style="min-height: 100dvh;"
>
  <form>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <input
            autocomplete="off"
            class="c4"
            data-testid="amount"
            id="amount-id"
            max="0.00000100000000000000"
            min="0"
            name="amount"
            placeholder="common.amount"
            required=""
            step="any"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="c5"
      />
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          <button
            class="c8"
            type="submit"
          >
            account.reclaimEscrow.reclaim
          </button>
        </div>
        <div
          class="c9"
        />
        <div
          class="c7"
        >
          <button
            class="c10"
            type="button"
          >
            account.reclaimEscrow.reclaimAll
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
